### PR TITLE
bpo-40747: Make py_version_nodot 3_10 not 310

### LINKED ISSUE #308: Python 3.10: _version_nodot, interpreter_version disagree on how to format it
There is a problem building and using wheels for python 3.10 xref pypa/wheel#354 and pypa/pip#8312. I think the source of the problem is that wheel's `bdist_wheel` calls `tags.interpreter_version`. But `tags.interpreter_version` seems to be producing `310`, which it will take from `sysconfig.get_config_var("py_version_nodot")` (if it exists), and there is no check that is conformant with the `3_10` format.

Internally, packaging uses `tags._version_nodot` which will always produce `3_10`, this is the "correct" value used by pip and in `tags.sys_tags`. 

[This is the code](https://github.com/python/cpython/blob/master/Lib/sysconfig.py#L89) that generates [that value](https://github.com/python/cpython/blob/master/Lib/sysconfig.py#L534) for CPython master. Note there is no '_'

### PR DESCRIPTION
**Pull request by mattip**
Fixes pypa/packaging#308, pypa/wheel#354, and maybe pypa/pip#8312.

The current value of `310` is ambiguous so use `3_10` instead. While this may fix `pip` and `wheel.bdist_wheel`, it has the potential to break other projects that depend on `_PY_VERSION_SHORT_NO_DOT` being 2 characters or that calculate the value via `'%d%d' % sys.version_info[:2]` themselves.

https://bugs.python.org/issue40747

### CODE REVIEW

**Comment by pablogsal** [comment]
Hmmm, this change may have some other unforeseen consequences. For instance, IIRC this affects the path (indeed, there is a test failing for Windows). Someone more familiar with packaging should take a look as well to make sure this does not have some other unwanted consequences. @brettcannon

**Comment by ned-deily** [comment]
Should probably have @zooba take a look as well.  I don't see where the short no dot variable from either copy of sysconfig is used in Unix-y builds of cpython.  I've done a quick macOS framework style build (--enable-framework) and I haven't noticed any obvious change in paths or any test suite failures. I'd be inclined to not make the change though unless it is likely to cause problems for downstream packages that do use the variable.

But is it really ambiguous?  cpython versions are always of the form 3.x, never 3.x.y, as far as constructing path names in builds, AFAIK.  That is, for unix-y builds using autoconf. @zooba could speak for the Windows build system.

**Comment by brettcannon** [comment]
It's ambiguous from the perspective of what is `"310"`: `3.10` or `31.0`? While we don't exactly plan on changing our numbering scheme right now, this is an ambiguity. Plus it just looks odd to me, which does somewhat matter as this will show up in wheel file names.

**Comment by mattip** [comment]
The [FAQ for PEP 425](https://www.python.org/dev/peps/pep-0425/#faq) has this:
> Why isn't there a . in the Python version number?
> CPython has lasted 20+ years without a 3-digit major release. This should continue for some time. Other implementations may use _ as a delimeter, since both - and . delimit the surrounding filename.

I think  "major release" in this context means 3.4, 3.5, 3.6, ... so we are about to hit a 3-digit major release.

**Comment by mattip** [comment]
Tests pass, including one new test for consistency.

File: configure.ac
Comment (ned-deily) [inline configure.ac]: On macOS, at least, 

```
$  ./configure --with-pydebug
[...]
checking SOABI... cpython-3_10d-darwin
usage: tr [-Ccsu] string1 string2
       tr [-Ccu] -d string1
       tr [-Ccu] -s string1
       tr [-Ccu] -ds string1 string2
checking LDVERSION... $(VERSION)$(ABIFLAGS)
[...]

```

**Review by ned-deily — CHANGES_REQUESTED** [review]

File: configure.ac
Comment (mattip) [reply configure.ac]: Huh. Strange it passes CI. Is the change on line 4740 also problematic? I cannot reproduce, could you try removing the `"` so that the `"." "_"` becomes `. _`

Comment (ned-deily) [reply configure.ac]: The `tr` error doesn't cause the configure run to fail.  It happens on the Linux CI runs, too: from the Github Ubuntu test:

```
676 checking SOABI... cpython-3_10d-x86_64-linux-gnu
677 tr: extra operand '_'
678 Only one string may be given when deleting without squeezing repeats.
679 Try 'tr --help' for more information.
680 checking LDVERSION... $(VERSION)$(ABIFLAGS)
```
(You don't see it in the log for the most recent Travis CI run because it appears that the build environment for the PR is cached and it didn't rerun configure.)

Comment (mattip) [reply configure.ac]: fixed by removing the -d from the tr command. Checked it by rerunning `autoconf` to recreate `configure` from `configure.am`.

**Comment by mattip** [comment]
I have made the requested changes; please review again

**Review by ned-deily — COMMENTED** [review]
Thanks, that fixes the `tr` errors.  Still should have a review by @zooba for Windows impact.

**Review by ned-deily — APPROVED** [review]

File: Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst
Comment (brettcannon) [inline Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst:1]:
```suggestion
The PEP 425 python tag, taken from ``py_version_nodot``, adds a ``_`` so ``cp310`` is now ``cp3_10``.
Patch by Matti Picus.
```

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
I think we are just waiting on @zooba at this point.

**Comment by multibuilder** [comment]
There is a merge conflict. I could merge or rebase to clear it, what is the preferred workflow?

**Comment by mattip** [comment]
There is a merge conflict, is it preferable to merge master or rebase master?

**Comment by brettcannon** [comment]
@mattip we do squash merges, so you can do a merge if that's easiest.

File: Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst
Comment (mattip) [reply Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst:1]: None of the other blurbs in that directory have a contributor name.

**Comment by mattip** [comment]
CI is passing. I added treatement of windows path and dlls, but did not check this works with, for instance, building and using a NumPy wheel.

File: PC/bdist_wininst/install.c
Comment (zooba) [inline PC/bdist_wininst/install.c:1689]: This will cause us to have to manually rebuild and commit `wininst-...exe` in Lib/distutils/command, and just glancing at the code I can see in the PR this appears to have a list of Python installs, so we may need logic here to only insert the underscore for 3.10 and later (or we can get PEP 632 approved and then ignore this file completely)

File: PCbuild/python.props
Comment (zooba) [inline PCbuild/python.props:205]: There are going to be a number of similar changes in Tools/msi

Comment (zooba) [reply PCbuild/python.props:205]: Likely some under Tools/nuget as well, though they may all be under Tools/msi or PC/layout.

**Review by zooba — CHANGES_REQUESTED** [review]
I'd expect to see changes under Tools/msi (once you add any, CI will run over that section) and probably PC/layout. The default install directory should probably also change to have the underscore.

Someone will need to rebuild the bdist_wininst stub, and possibly add more logic if it does actually offer a choice of install targets (or someone will need to approve PEP 632).

I'd expect a significant What's New section and a few documentation updates. Pretty sure the "python310.zip" entry on sys.path is inferred from the DLL name on Windows, and the python310._pth file certainly is. I only recently touched the user site packages directory, so I think that's fine with your change here.

_Alternatively_, we don't really need to do this, since we're not even planning to get to Python 4.x, let alone 31.x ;) Seems like plenty of time to come up with alternative file layouts.

**Comment by mattip** [comment]
> Alternatively, we don't really need to do this, since we're not even planning to get to Python 4.x, let alone 31.x ;) Seems like plenty of time to come up with alternative file layouts.

I am fine with closing this, but then pypa/packaging should align itself with that approach.

File: PC/bdist_wininst/install.c
Comment (mattip) [reply PC/bdist_wininst/install.c:1689]: Even if PEP 632 to remove distutils is approved, these changes will need to be reflected in pypa/setuptools' copy of distutils, or did they drop bdist_wininst ?

**Comment by mattip** [comment]
I searched Doc, Tools/msi, and PC/layout for `nodot` and (case insensitive) `major` and added fixes. The only change in Doc was the outdated directions for linking in windows, which did not mention the linking pragma in `pyconfig.h`.

File: PCbuild/python.props
Comment (mattip) [reply PCbuild/python.props:205]: grepped and changed where needed (I think)

**Comment by mattip** [comment]
somewhat surprisingly, CI passed with the MSI changes

**Comment by pablogsal** [comment]
Unfortunately, this issue is also blocking running the `pyperformance` suite with master. @zooba, as you requested changes, how would you like to proceed?

File: Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst
Comment (brettcannon) [reply Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst:1]: That doesn't mean the other news entries are wrong. 😉 (And this is coming from the person who created this workflow, so I can say with certainty you can put your name in the news entry.)

**Comment by brettcannon** [comment]
We should aim to resolve this by the end of the core dev sprint so we can unblock the rest of the ecosystem.

**Comment by hugovk** [comment]
> We should aim to resolve this by the end of the core dev sprint so we can unblock the rest of the ecosystem.

For those of us who didn't know, it's [next week](https://python-core-sprint-2020.readthedocs.io/):

> The sprint is scheduled for Oct 19 - 23, 2020. This will be an online event instead of the in-person sprint.

Sounds good! Should it be added to https://python-core-sprint-2020.readthedocs.io/projects.html?

File: Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst
Comment (mattip) [reply Misc/NEWS.d/next/Build/2020-05-25-09-46-32.bpo-40747.ef15ix.rst:1]: I would prefer not to, but won't stop you from commiting the change.

**Comment by brettcannon** [comment]
@hugovk nah, it will just be something that some of us go and discuss on the side.

**Comment by brettcannon** [comment]
@hugovk I decided not to risk forgetting, so I went ahead and added it to the TODO list. Thanks for the idea!

**Comment by brandtbucher** [comment]
@pablogsal downgrading to `pyperformance==1.0.0` is a temporary workaround.

**Comment by pablogsal** [comment]
> @pablogsal downgrading to `pyperformance==1.0.0` is a temporary workaround.

Thanks @brandtbucher ! Unfortunately, we were working on the benchmarks for the speed.python.org server and downgrading uses a different versions of the packages and therefore the generated data cannot be compared with the previous data 🙁

**Comment by brettcannon** [comment]
For fellow core devs: feel free to add/remove yourself from https://python-core-sprint-2020.readthedocs.io/projects.html#wheel-interpreter-naming-for-3-10 and I will try to schedule a meeting when we can chat about this. I will wait until Monday to schedule based on who does (not) want to attend to make it timezone friendly (e.g. Pablo I know is London and I'm Vancouver).

**Comment by pablogsal** [comment]
> I will wait until Monday to schedule based on who does (not) want to attend to make it timezone friendly (e.g. Pablo I know is London and I'm Vancouver).

My sleep schedule is already destroyed and I am a night person, so feel free to push the meeting closer to your timezone if needed 😉

File: PC/bdist_wininst/install.c
Comment (zooba) [reply PC/bdist_wininst/install.c:1689]: Question for @jaraco, but hopefully they dropped it.

File: Tools/msi/doc/doc.wixproj
Comment (zooba) [inline Tools/msi/doc/doc.wixproj:13]: We should probably separate the minor and micro version of the doc filename too - `python3_10_0.chm` rather than `python3_100.chm`. The actual change ought to be somewhere under Doc, IIRC.

File: PC/bdist_wininst/install.c
Comment (jaraco) [reply PC/bdist_wininst/install.c:1689]: Thusfar, it only [raises a DeprecationWarning](https://github.com/pypa/distutils/blob/68f5b8181e1f42fd12586752d93e97bda1649d1d/distutils/command/bdist_wininst.py#L65-L67) (aka not a user-visible warning), so may need some care to disable it.

**Comment by ned-deily** [comment]
FYI, I've built and run with the PR merged up to the current HEAD of master with both a vanilla installed Linux config and my standard macOS installer build.  In both cases, no test failures due to the PR were noted.  In both cases, I also ran `python3.10 -m test.pythoninfo` and searched the output for the substring `310`.  I found two instances of that:

1. the zipfile path (which @zooba noted above).  On Linux with an install prefix of /tmp/b:
`sys.path: ['/tmp/a', '/tmp/b/lib/python310.zip', '/tmp/b/lib/python3.10', '/tmp/b/lib/python3.10/lib-dynload', '/home/nad/.local/lib/python3.10/site-packages', '/tmp/b/lib/python3.10/site-packages']`

2. the `cache_tag` attribute of `sys.implementation`:
`sys.implementation: namespace(name='cpython', cache_tag='cpython-310', version=sys.version_info(major=3, minor=10, micro=0, releaselevel='alpha', serial=1), hexversion=50987169, _multiarch='i386-linux-gnu')
`

I have no opinion ATM of the importance of either one, just reporting them.

File: PC/bdist_wininst/install.c
Comment (zooba) [reply PC/bdist_wininst/install.c:1689]: I think it's fine to merge with or without this change, and we'll just track the distutils breakage as a new bug. Don't hold up this PR on it.

**Comment by brettcannon** [comment]
@mattip we discussed this at the sprint and decided that once Steve's concerns are addressed and I write a quick PEP this week on the chnage we can merge the PR.

**Comment by zooba** [comment]
The docs filename change seems to be at line 108 of https://github.com/mattip/cpython/blob/py_version_nodot/Doc/conf.py

Turns out the MSI build in PR is using the "friendly" option that doesn't fail when the doc file is missing, but [this line](https://github.com/python/cpython/pull/20333/checks?check_run_id=1246054173#step:3:1860) shows that it wasn't found.

I think we should be fine to just allow the extra dots in the .chm name though, rather than going to underscores. And actually, I think we should try going to dots in the DLL name as well. Windows should be able to handle it, and it's more consistent between `libpython3.10.so` and `python3.10.dll` than having the underscore in there.

And if we're going to cause (a few) people to have to update their code anyway, we may as well end up somewhere more consistent :)

**Comment by mattip** [comment]
I'm not sure I understand what still needs to be changed. Could a reviewer who does understand push changes to this PR?

**Comment by zooba** [comment]
I can push changes, just didn't want to intrude since you've been active.

File: configure.ac
Comment (merwok) [reply configure.ac]: (FWIW the error makes sense: tr is either `original replacement` or `-d original` to delete)

**Comment by mattip** [comment]
@zooba or anyone else, please feel free to make any necessary changes to move this forward

**Comment by zooba** [comment]
I've had to reopen as #22858 with a branch on my own fork, since I can only edit mattip's through the web UI (which, as you'll see, was not going to be worthwhile).

**Pull request closed without merging**
